ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,866 @@
+
+    def cmd_mkdir(self, path):
+        self._log(f"Make directory '{path}'")
+
+        os.makedirs(path, exist_ok=True)  # we allow that the directory already exists
+
+    def cmd_rmdir(self, path):
+        self._log(f"Remove directory '{path}'")
+
+        import shutil
+
+        shutil.rmtree(path)
+
+    def cmd_exists(self, path):
+        self._log(f"Check if path exists '{path}'")
+
+        if os.path.exists(path):
+            print("__exists=1__")
+        else:
+            print("__exists=0__")
+
+    def cmd_remove(self, path):
+        self._log(f"Remove file '{path}'")
+
+        # if self.debugging == True: # make file backup to retrieve it later
+        # import shutil
+        # shutil.copyfile(path, path+".debug_backup")
+
+        os.remove(path)
+
+
+class ShellCommandSend:
+    """
+    Helper class for sending shell commands in generic and OS independent form
+
+    The class is designed to send shell commands to an ssh connection. Nevertheless, it can be
+    used for send commands to different local shell as well, which is useful for testing. Since
+    the concept uses the ipyparallel package (class ipyparallel.cluster.ShellCommandReceive) for
+    executing the commands, it is necessary that a valid python installation (python_path) is provided.
+    By calling has_python this can be evaluated. Furthermore, get_shell_info can be called to retrieve OS and
+    shell information independent of a valid python installation.
+
+    Since some operation require different handling for different platforms/shells the class gathers
+    necessary information during initialization. This per default automatically at object instantiation.
+    Nevertheless, it can be postponed (initialize=False) but before any member function is called, the
+    initialize function has to be called.
+
+    Beside generic check_output[...] functions (equivalent to subprocess.check_output), the class provides
+    specific shell commands which have a cmd_ prefix. When adding new functions make sure that an
+    equivalent is added in the ShellCommandReceive class as well.
+
+    To start processes through an ssh connection on a windows server that stays alife after the ssh connection
+    is closed, the process must be started with the breakaway creation flag set. Which works fine on 'normal'
+    machine and VMs is denied on windows github runners (I guess because of security reasons). Hence, it was
+    necessary to implement a work-a-round to enable CI in github. In case breakaway is not supported, a detached
+    ssh connection is started (by the ShellCommandSend object) which stays open until the process to start has
+    finished. see _cmd_send for further details.
+    """
+
+    package_name = "ipyparallel.shellcmd"  # package name for send the command
+    output_template = re.compile(
+        r"__([a-z][a-z0-9_]+)=([a-z0-9\-\.]+)__", re.IGNORECASE
+    )
+    receiver_code = inspect.getsource(ShellCommandReceive)
```

Comment:
rather than extracting snippets, can we make it a whole standalone file?

Reply:
see dicussion [above](https://github.com/ipython/ipyparallel/pull/842#discussion_r1663862160)